libopenage/util/csv.h

```diff
@@ -0,0 +1,160 @@
+// Copyright 2013-2017 the openage authors. See copying.md for legal info.
+
+#pragma once
+
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "../error/error.h"
+
+#include "compiler.h"
+#include "dir.h"
+
+namespace openage {
+namespace util {
+
+
+/**
+ * Type for storing csv data:
+ * {filename: [line, ...]}.
+ */
+using csv_file_map_t = std::unordered_map<std::string, std::vector<std::string>>;
+
+
+/**
+ * Load a multi csv file into a csv_file_map_t
+ */
+csv_file_map_t load_multi_csv_file(Dir basedir, const std::string &fname);
+
+
+/**
+ * read a single csv file.
+ * call the destination struct .fill() method for actually storing line data
+ */
+template<typename lineformat>
+void read_csv_file(const std::string &fname, std::vector<lineformat> &out, csv_file_map_t *file_map=nullptr) {
+	size_t line_count = 0;
+	lineformat current_line_data;
+	std::vector<char> strbuf;
+
+	// instead of reading from the file, use the file_map as buffer.
+	if (file_map && file_map->count(fname)) {
+		std::vector<std::string> lines = file_map->at(fname);
+
+		for (auto &line : lines) {
+			line_count += 1;
+
+			// create writable copy, for tokenisation
+			if (unlikely(strbuf.size() <= line.size())) {
+				strbuf.resize(line.size() + 1);
+			}
+			memcpy(strbuf.data(), line.c_str(), line.size() + 1);
+
+			// use the line copy to fill the current line struct.
+			int error_column = current_line_data.fill(strbuf.data());
```

Comment:
Ah now I see what you mean... It is finally consumed [here](https://github.com/SFTtech/openage/blob/master/libopenage/util/strings.cpp#L90) - and there write access is required - which is still not completely broken - if the string is not modified in length one can use ```&line[0]````- and continue to avoid any copy